core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -525,21 +524,22 @@ private void applyNewOrUpdatedShards(final ClusterChangedEvent event) {
                 assert currentRoutingEntry.isSameAllocation(shardRouting) :
                     "local shard has a different allocation id but wasn't cleaning by applyDeletedShards. "
                         + "cluster state: " + shardRouting + " local: " + currentRoutingEntry;
-                if (isPeerRecovery(shardRouting)) {
-                    final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(routingTable, nodes, shardRouting);
-                    // check if there is an existing recovery going, and if so, and the source node is not the same, cancel the recovery to restart it
-                    if (recoveryTargetService.cancelRecoveriesForShard(indexShard.shardId(), "recovery source node changed", status -> !status.sourceNode().equals(sourceNode))) {
-                        logger.debug("[{}][{}] removing shard (recovery source changed), current [{}], global [{}])", shardRouting.index(), shardRouting.id(), currentRoutingEntry, shardRouting);
-                        // closing the shard will also cancel any ongoing recovery.
-                        indexService.removeShard(shardRouting.id(), "removing shard (recovery source node changed)");
-                        shardHasBeenRemoved = true;
+                if (shardRouting.isPeerRecovery()) {
+                    RecoveryState recoveryState = indexShard.recoveryState();
+                    if (recoveryState != null && recoveryState.getStage() != RecoveryState.Stage.DONE) { // shard is still recovering
+                        if (recoveryState.getType() == RecoveryState.Type.PRIMARY_RELOCATION || recoveryState.getType() == RecoveryState.Type.REPLICA) {
+                            final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
+                            if (recoveryState.getSourceNode().equals(sourceNode) == false) {
```

Comment:
good idea. I've taken the cancelRecoveriesForShard approach.

>  I wonder if we should also catch the case where a primary relocates on the master and not have to worry about all of this here. I think it will be simpler all in all.

I'm less convinced of that one. Fully handling it on the master means that a shard that is in POST_RECOVERY can be cancelled just because primary relocation completed before shard was activated by the master node. This is especially bad if there is some load on master (with longer queue of tasks) where shards take a bit of time to get started and relocation of primary is going on. I think that it is sufficient for master to determine the recovery source for each shard. We can then easily check if it needs to be changed.